File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Map<Character, Character> mappings = new HashMap<>();
        mappings.put(')','(');
        mappings.put('}','{');
        mappings.put(']','[');

        Stack<Character> stack = new Stack<>();
        for(char c: s.toCharArray()){
            if(mappings.containsKey(c)){
                if(!stack.isEmpty() && stack.peek() == mappings.get(c)){
                    stack.pop();
                } else {
                    return false;
                }
            }else{
                stack.push(c);
            }
        }

        return stack.isEmpty();
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        d = {
            ']' : '[',
            ')' : '(',
            '}' : '{'
        }
        stack = []
        for ch in s:
            if ch in stack:
                if stack and stack[-1] == d[ch]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(ch)
        
        print(stack)
        return len(stack) == 0



